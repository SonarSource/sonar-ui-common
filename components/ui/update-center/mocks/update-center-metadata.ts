/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { MetaDataInformation, MetaDataVersionInformation } from '../update-center-metadata';

export function mockMetaDataVersionInformation(
  overrides?: Partial<MetaDataVersionInformation>
): MetaDataVersionInformation {
  return {
    version: '5.13',
    date: '2019-05-31',
    compatibility: '6.7',
    archived: false,
    downloadURL: 'https://example.com/sonar-java-plugin-5.13.0.18197.jar',
    changeLogUrl: 'https://example.com/sonar-java-plugin/release',
    ...overrides
  };
}

export function mockMetaDataInformation(
  overrides?: Partial<MetaDataInformation>
): MetaDataInformation {
  return {
    name: 'SonarJava',
    key: 'java',
    isSonarSourceCommercial: true,
    organization: {
      name: 'SonarSource',
      url: 'http://www.sonarsource.com/'
    },
    category: 'Languages',
    license: 'SonarSource',
    issueTrackerURL: 'https://jira.sonarsource.com/browse/SONARJAVA',
    sourcesURL: 'https://github.com/SonarSource/sonar-java',
    versions: [
      mockMetaDataVersionInformation({ version: '2.0' }),
      mockMetaDataVersionInformation({ version: '1.0', archived: true })
    ],
    ...overrides
  };
}
