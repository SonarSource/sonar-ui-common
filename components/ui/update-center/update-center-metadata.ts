/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export interface MetaDataInformation {
  category?: string;
  isSonarSourceCommercial?: boolean;
  issueTrackerURL?: string;
  key?: string;
  license?: string;
  name: string;
  organization?: {
    name: string;
    url?: string;
  };
  sourcesURL?: string;
  versions?: MetaDataVersionInformation[];
}

export interface MetaDataVersionInformation {
  archived?: boolean;
  changeLogUrl?: string;
  compatibility?: string;
  date?: string;
  description?: string;
  downloadURL?: string | AdvancedDownloadUrl[];
  version: string;
}

export interface AdvancedDownloadUrl {
  label?: string;
  url: string;
}
