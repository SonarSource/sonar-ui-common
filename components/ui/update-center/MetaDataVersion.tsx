/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import classNames from 'classnames';
import * as React from 'react';
import { AdvancedDownloadUrl, MetaDataVersionInformation } from './update-center-metadata';

export interface MetaDataVersionProps {
  versionInformation: MetaDataVersionInformation;
}

export default function MetaDataVersion(props: MetaDataVersionProps) {
  const {
    versionInformation: {
      archived,
      changeLogUrl,
      compatibility,
      date,
      description,
      downloadURL,
      version
    }
  } = props;

  const fallbackLabel = 'Download';

  const advancedDownloadUrls = isAdvancedDownloadUrlArray(downloadURL)
    ? downloadURL.map(url => ({ ...url, label: url.label || fallbackLabel }))
    : [{ label: fallbackLabel, url: downloadURL }];

  return (
    <div
      className={classNames('update-center-meta-data-version', {
        'update-center-meta-data-version-archived': archived
      })}>
      <div className="update-center-meta-data-version-version">{version}</div>

      <div className="update-center-meta-data-version-release-info">
        {date && <time className="update-center-meta-data-version-release-date">{date}</time>}

        {compatibility && (
          <span className="update-center-meta-data-version-compatibility">{compatibility}</span>
        )}
      </div>

      {description && (
        <div className="update-center-meta-data-version-release-description">{description}</div>
      )}

      {(advancedDownloadUrls.length > 0 || changeLogUrl) && (
        <div className="update-center-meta-data-version-release-links">
          {advancedDownloadUrls.length > 0 &&
            advancedDownloadUrls.map(
              (advancedDownloadUrl, i) =>
                advancedDownloadUrl.url && (
                  // eslint-disable-next-line react/no-array-index-key
                  <span className="update-center-meta-data-version-download" key={i}>
                    <a href={advancedDownloadUrl.url} rel="noopener noreferrer" target="_blank">
                      {advancedDownloadUrl.label}
                    </a>
                  </span>
                )
            )}

          {changeLogUrl && (
            <span className="update-center-meta-data-version-release-notes">
              <a href={changeLogUrl} rel="noopener noreferrer" target="_blank">
                Release notes
              </a>
            </span>
          )}
        </div>
      )}
    </div>
  );
}

function isAdvancedDownloadUrlArray(
  downloadUrl: string | AdvancedDownloadUrl[] | undefined
): downloadUrl is AdvancedDownloadUrl[] {
  return !!downloadUrl && typeof downloadUrl !== 'string';
}
