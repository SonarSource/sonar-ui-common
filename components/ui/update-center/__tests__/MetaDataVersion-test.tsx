/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import MetaDataVersion, { MetaDataVersionProps } from '../MetaDataVersion';
import { mockMetaDataVersionInformation } from '../mocks/update-center-metadata';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
  expect(
    shallowRender({
      versionInformation: mockMetaDataVersionInformation({
        downloadURL: [{ label: 'macos 64 bits', url: '' }],
      }),
    })
  ).toMatchSnapshot('with advanced downloadUrl');
  expect(
    shallowRender({
      versionInformation: { version: '2.0' },
    })
  ).toMatchSnapshot('with very few info');
});

function shallowRender(props?: Partial<MetaDataVersionProps>) {
  return shallow(
    <MetaDataVersion versionInformation={mockMetaDataVersionInformation()} {...props} />
  );
}
