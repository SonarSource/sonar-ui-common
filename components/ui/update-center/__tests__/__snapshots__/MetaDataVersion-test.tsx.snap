// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="update-center-meta-data-version"
>
  <div
    className="update-center-meta-data-version-version"
  >
    5.13
  </div>
  <div
    className="update-center-meta-data-version-release-info"
  >
    <time
      className="update-center-meta-data-version-release-date"
    >
      2019-05-31
    </time>
    <span
      className="update-center-meta-data-version-compatibility"
    >
      6.7
    </span>
  </div>
  <div
    className="update-center-meta-data-version-release-links"
  >
    <span
      className="update-center-meta-data-version-download"
      key="0"
    >
      <a
        href="https://example.com/sonar-java-plugin-5.13.0.18197.jar"
        rel="noopener noreferrer"
        target="_blank"
      >
        Download
      </a>
    </span>
    <span
      className="update-center-meta-data-version-release-notes"
    >
      <a
        href="https://example.com/sonar-java-plugin/release"
        rel="noopener noreferrer"
        target="_blank"
      >
        Release notes
      </a>
    </span>
  </div>
</div>
`;

exports[`should render correctly: with advanced downloadUrl 1`] = `
<div
  className="update-center-meta-data-version"
>
  <div
    className="update-center-meta-data-version-version"
  >
    5.13
  </div>
  <div
    className="update-center-meta-data-version-release-info"
  >
    <time
      className="update-center-meta-data-version-release-date"
    >
      2019-05-31
    </time>
    <span
      className="update-center-meta-data-version-compatibility"
    >
      6.7
    </span>
  </div>
  <div
    className="update-center-meta-data-version-release-links"
  >
    <span
      className="update-center-meta-data-version-release-notes"
    >
      <a
        href="https://example.com/sonar-java-plugin/release"
        rel="noopener noreferrer"
        target="_blank"
      >
        Release notes
      </a>
    </span>
  </div>
</div>
`;

exports[`should render correctly: with very few info 1`] = `
<div
  className="update-center-meta-data-version"
>
  <div
    className="update-center-meta-data-version-version"
  >
    2.0
  </div>
  <div
    className="update-center-meta-data-version-release-info"
  />
  <div
    className="update-center-meta-data-version-release-links"
  />
</div>
`;
